### DISCUSSION TITLE
Limiting factors for decreasing the distance between hook and carrot.

### DISCUSSION THREAD

**Discussion by AndyPermaRobotics** [zauberzeug/rosys#41] [Ideas / Feature Requests]
I made some experiments with different Hook/Carrot Distances and discovered, that the smaller the carrot to hook distance is, the better the robot drives the path especially for curves:
![telegram-cloud-photo-size-4-5857354425964806242-y](https://github.com/zauberzeug/rosys/assets/129860214/e89d854b-7037-4048-83bf-0f83714411a7)

So in which scenarios should we choose a bigger carrot_distance?

Currently the only reason that I could find, seems to be a problem with the carrot.move() method in this line:
`while hook.distance(self.offset_point) < distance:`
The carrot is moved until the configured distance between carrot and hook is reached again.
But if the robot is driving faster, the drivers update interval of 100ms can be enough for the robot to drive two times the distance between hook and carrot and in that case the move() condition of the carrot is fulfilled, but the other way around.

<img width="885" alt="image" src="https://github.com/zauberzeug/rosys/assets/129860214/eaba7f6f-9224-4358-b0a3-02ace23b222b">

This results in the robot switches between driving forward and backward in small intervals:
<img width="420" alt="image" src="https://github.com/zauberzeug/rosys/assets/129860214/30809f8d-ab39-4027-ad62-438edd085a4f">

This is the only limiting factor for decreasing the distance between carrot and hook that I could find.
Is there any other?

If not, I would suggest to adjust the logic in the carrot.move() method.
I can also try to find a solution and may create a pull request.
But before that, I want to make sure, that it makes sense.

**Comment by falkoschindler** [comment]
Hi Andreas, that's a great question!

It took me some time to wrap my head around this concept and the decisions that lead to its implementation a while ago. But I think I remember:
One aspect is to consider noisy position data. In a real-world scenario the robot is never exactly where it is supposed to be. Either from absolute position sensors like GPS or outside-in camera tracking or from inaccurate wheel odometry, the robot pose might jump from time to time. If you choose a too small distance between carrot and hook, the vector from hook to carrot will always point orthogonally towards the path and the robot will be busy oscillating around the line but not driving forward. A larger carrot-hook distance allows for some deviation and results in forward motion as long as the hook is "close enough" to the line. So this parameter serves as a kind of filter and controls how strictly the robot follows the path.

Does that make sense?

**Reply by AndyPermaRobotics** [reply]
Hi Falko,

thanks for you explanation. That makes sense.
Still it's a pity, that the greater carrot-hook distance leads to a less accurate driving of the splines.
I will think about it. Maybe I can find a solution that mitigates the problems.
But now I know, that the signal noice is a crucial aspect. Thanks!
